corehq/apps/domain/decorators.py

```diff
@@ -474,7 +474,14 @@ def _inner(request, domain, *args, **kwargs):
     return _outer
 
 
-def _two_factor_required(view_func, domain, couch_user):
+def _two_factor_required(view_func, domain, couch_user, request):
```

Comment:
Nit: now that `request` is a required argument here, and the callers of this function get `domain` and `couch_user` from the request, those arguments could be removed, which makes the signature nicer
```suggestion
def _two_factor_required(view_func, request):
    domain = request.project
    couch_user = request.couch_user
```

Reply:
yes, i certainly thought about that, but https://github.com/dimagi/commcare-hq/blob/715ccb00f4ecfca594614815fb4382cae305042f/corehq/apps/domain/decorators.py#L452

Reply:
I see. Maybe I overlooked something, but the (non-test) code I saw either called it like
```py
_two_factor_required(view_fn, request.project, request.couch_user, request)
```

or seemed to have earlier code that used those attributes. I could be wrong though. Maybe they (at least `request.project`) are not set for this one?

https://github.com/dimagi/commcare-hq/blob/715ccb00f4ecfca594614815fb4382cae305042f/corehq/apps/domain/decorators.py#L446-L452

Aside: I'm just learning in this PR that `reqeust.project` is a thing. That's obviously related by different from `request.domain` which I did know about.

Reply:
do you think it's safer to ensure that `request.project` and `request.couch_user` are set here then? 
https://github.com/dimagi/commcare-hq/blob/715ccb00f4ecfca594614815fb4382cae305042f/corehq/apps/domain/decorators.py#L446-L452

I thought passing request into the two places `_two_factor_required()` is used (outside of tests) felt safer than making that change to `request` but I'm open to changing my mind here

Reply:
I don't know where `request.project` is set, but was assuming it was _always_ set before this code is run. It would be ideal to only set it in one place and all code that uses it knows it is safe to reference. It seems less than ideal to add more/new code here that sets `request.project` if we do not know (a) why it isn't already set and (b) that setting it here will satisfy requirements of all other code that references it.

If we don't have good solid answers to those questions then your code is probably safer, although I think it is a bit harder to reason about.

Reply:
In any case, it appears that `request.couch_user` is always set, so no need to pass it in separately from `request`.

Reply:
ok I tidied things up a bit in the new commits. it looks like `_two_factor_required` is used in the `two_factor_check` decorator which is called by the api-related decorators, none of which I feel are guaranteed to populate `request.project` and I'm not super keen to populate `request.project` in this decorator as I'm not 100% sure how that might effect things downstream. I opted to consolidated `request.couch_user` and `request` and also made the signature more explicit in what it's asking for and added some docs.

Reply:
Without having looked closely at the usage here, I may be able to provide some context, as the state of our auth code is quite messy and brittle.  `request.project` refers to the domain object, whereas `request.domain` refers to the string.  The former is set here (in a function that both mutates an input argument and returns a value), which is called inside `login_and_domain_required` - a root decorator for most domain-specific views.
https://github.com/dimagi/commcare-hq/blob/331dd894b1d6e88e550386e3e9710ca375641d27/corehq/apps/domain/decorators.py#L70-L74
Unless you're in a report, in which case it's set in the dispatch method.

`request.domain` is set in [the users middleware](https://github.com/dimagi/commcare-hq/blob/331dd894b1d6e88e550386e3e9710ca375641d27/corehq/apps/users/middleware.py#L37-L40), which also sets `request.couch_user`

It's also tough to say exactly in which circumstances any of these things will be present.  Decorators can be called in different orders, so things set in one might be available to other decorators in some views, but not in others.  Middleware is called in a consistent order (and before decorators), but with some authentication methods, authentication happens and the user is set in a _decorator_, not middleware, which can make other middleware or decorators maybe not see the user as logged in, depending on how the user provided authentication and the order of decorators.

It's also the case that some code ([scheduled reports](https://github.com/dimagi/commcare-hq/blob/331dd894b1d6e88e550386e3e9710ca375641d27/corehq/apps/reports/generic.py#L209-L214), at least) runs view code with a fake request where some of these fields are set manually, and which I believe doesn't go through middleware.

Overall, it's pretty difficult to assume any sort of tight constraints on the current state of a request object or authentication pathway.